File: sdk/npm/src/providers/temporam.ts
import { randomInt } from 'crypto';
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'temporam';
const ORIGIN = 'https://temporam.com';

const JSON_HEADERS = {
  Accept: 'application/json',
  'Accept-Encoding': 'gzip, deflate, br',
  'User-Agent':
    'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36',
};

type DomainItem = {
  domain?: string;
};

type DomainsResponse = {
  data?: DomainItem[];
};

type EmailsResponse = {
  data?: any[];
};

type EmailDetailResponse = {
  data?: any;
};

function randomLocal(length = 18): string {
  const chars = 'abcdefghijklmnopqrstuvwxyz0123456789';
  let out = 'sdk';
  for (let i = 0; i < length; i++) {
    out += chars[randomInt(chars.length)];
  }
  return out;
}

async function getJson<T>(path: string): Promise<T> {
  const response = await fetchWithTimeout(`${ORIGIN}${path}`, {
    headers: JSON_HEADERS,
  });
  if (!response.ok) {
    throw new Error(`temporam request failed: ${response.status}`);
  }
  return await response.json() as T;
}

async function listDomains(): Promise<string[]> {
  const data = await getJson<DomainsResponse>('/api/domains');
  const domains = (data.data ?? [])
    .map(item => (item.domain ?? '').trim())
    .filter(Boolean);
  if (domains.length === 0) {
    throw new Error('temporam: domain list is empty');
  }
  return domains;
}

function normalizeTemporamEmail(raw: any, email: string): Email {
  return normalizeEmail({
    ...raw,
    id: raw?.id || raw?.uuid || '',
    from: raw?.fromEmail || raw?.from || '',
    to: raw?.toEmail || raw?.to || email,
    date: raw?.createdAt || raw?.created_at || raw?.date || '',
  }, email);
}

export async function generateEmail(domain?: string | null): Promise<InternalEmailInfo> {
  const domains = await listDomains();
  const selectedDomain = domain
    ? domains.find(item => item === domain)
    : domains[randomInt(domains.length)];
  if (!selectedDomain) {
    throw new Error(`temporam: unsupported domain ${domain}`);
  }

  return {
    channel: CHANNEL,
    email: `${randomLocal()}@${selectedDomain}`,
  };
}

export async function getEmails(email: string): Promise<Email[]> {
  const list = await getJson<EmailsResponse>(`/api/emails?email=${encodeURIComponent(email)}&limit=50`);
  const rows = Array.isArray(list.data) ? list.data : [];
  const out: Email[] = [];

  for (const row of rows) {
    const id = row?.id || row?.uuid || '';
    if (!id) {
      out.push(normalizeTemporamEmail(row, email));
      continue;
    }

    try {
      const detail = await getJson<EmailDetailResponse>(`/api/emails/${encodeURIComponent(String(id))}`);
      out.push(normalizeTemporamEmail(detail.data || row, email));
    } catch {
      out.push(normalizeTemporamEmail(row, email));
    }
  }

  return out;
}

File: sdk/go/provider/temporam.go
package provider

import (
	"crypto/rand"
	"encoding/json"
	"fmt"
	"io"
	"math/big"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const temporamOrigin = "https://temporam.com"

type temporamDomainsResp struct {
	Data []struct {
		Domain string `json:"domain"`
	} `json:"data"`
}

type temporamEmailsResp struct {
	Data []map[string]any `json:"data"`
}

type temporamDetailResp struct {
	Data map[string]any `json:"data"`
}

func temporamJSONRequest(path string, out any) error {
	req, err := http.NewRequest("GET", temporamOrigin+path, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json")
	req.Header.Set("Accept-Encoding", "gzip, deflate, br")
	req.Header.Set("User-Agent", GetCurrentUA())

	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	if err := CheckHTTPStatus(resp, "temporam"); err != nil {
		return err
	}
	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if err := json.Unmarshal(body, out); err != nil {
		return fmt.Errorf("temporam: parse response: %w", err)
	}
	return nil
}

func temporamRandomInt(n int) int {
	if n <= 1 {
		return 0
	}
	v, err := rand.Int(rand.Reader, big.NewInt(int64(n)))
	if err != nil {
		return 0
	}
	return int(v.Int64())
}

func temporamRandomLocal() string {
	const chars = "abcdefghijklmnopqrstuvwxyz0123456789"
	var b strings.Builder
	b.WriteString("sdk")
	for i := 0; i < 18; i++ {
		b.WriteByte(chars[temporamRandomInt(len(chars))])
	}
	return b.String()
}

func temporamDomains() ([]string, error) {
	var data temporamDomainsResp
	if err := temporamJSONRequest("/api/domains", &data); err != nil {
		return nil, err
	}
	domains := make([]string, 0, len(data.Data))
	for _, item := range data.Data {
		domain := strings.TrimSpace(item.Domain)
		if domain != "" {
			domains = append(domains, domain)
		}
	}
	if len(domains) == 0 {
		return nil, fmt.Errorf("temporam: domain list is empty")
	}
	return domains, nil
}

func TemporamGenerate(domain *string) (*CreatedMailbox, error) {
	domains, err := temporamDomains()
	if err != nil {
		return nil, err
	}
	selected := domains[temporamRandomInt(len(domains))]
	if domain != nil && *domain != "" {
		selected = ""
		for _, item := range domains {
			if item == *domain {
				selected = item
				break
			}
		}
		if selected == "" {
			return nil, fmt.Errorf("temporam: unsupported domain %s", *domain)
		}
	}

	email := temporamRandomLocal() + "@" + selected
	return &CreatedMailbox{
		Channel: "temporam",
		Email:   email,
		Token:   email,
	}, nil
}

func temporamNormalize(raw map[string]any, recipient string) NormEmail {
	if raw["id"] == nil {
		raw["id"] = raw["uuid"]
	}
	if raw["from"] == nil {
		raw["from"] = raw["fromEmail"]
	}
	if raw["to"] == nil {
		raw["to"] = raw["toEmail"]
	}
	if raw["date"] == nil {
		raw["date"] = raw["createdAt"]
	}
	return NormalizeMap(raw, recipient)
}

func TemporamGetEmails(token string, email string) ([]NormEmail, error) {
	recipient := token
	if recipient == "" {
		recipient = email
	}
	if recipient == "" {
		return nil, fmt.Errorf("temporam: missing email")
	}

	var list temporamEmailsResp
	path := "/api/emails?email=" + url.QueryEscape(recipient) + "&limit=50"
	if err := temporamJSONRequest(path, &list); err != nil {
		return nil, err
	}

	out := make([]NormEmail, 0, len(list.Data))
	for _, row := range list.Data {
		raw := row
		if id, ok := row["id"].(string); ok && id != "" {
			var detail temporamDetailResp
			if err := temporamJSONRequest("/api/emails/"+url.PathEscape(id), &detail); err == nil && detail.Data != nil {
				raw = detail.Data
			}
		}
		out = append(out, temporamNormalize(raw, recipient))
	}
	return out, nil
}
